connman/vpn/vpn-provider.c

```diff
@@ -1713,8 +1713,19 @@ int __vpn_provider_connect(struct vpn_provider *provider, DBusMessage *msg)
 	} else
 		return -EOPNOTSUPP;
 
-	if (err == -EINPROGRESS)
-		vpn_provider_set_state(provider, VPN_PROVIDER_STATE_ASSOCIATION);
+	if (err == -EINPROGRESS) {
+		vpn_provider_set_state(provider,
+					VPN_PROVIDER_STATE_ASSOCIATION);
+
+		/*
+		 * If the VPN does not use VPN agent do direct transition to
+		 * connect in order to support the complete state machine.
+		 */
+		if (provider->driver && provider->driver->uses_vpn_agent &&
+				provider->driver->uses_vpn_agent(provider))
```

Comment:
Is there anywhere to record the semantics of `uses_vpn_agent`? I had a quick look but didn't see anywhere that looked appropriate, but maybe I missed it.

If I understand correctly, if I'm writing a vpn driver and define a `uses_vpn_agent` function, then a zero return value means "use VPN agent" and a non-zero return value means "skip VPN agent".

And presumably the idea is that this matches with the general connman semantics of "zero means success"?

Reply:
Zero or non-existing function means that VPN agent is used. But then again, using `bool` here would be self-explanatory. I'll change to that.

Reply:
I'm probably misunderstanding, but I would expect `uses_vpn_agent()` returning `true` to mean "stay on `VPN_PROVIDER_STATE_ASSOCIATION` and don't go directly on to `VPN_PROVIDER_STATE_CONNECT`".

If that's the intention, I think the function should be negated:
```suggestion
				!provider->driver->uses_vpn_agent(provider))
```
Sorry to dwell on this, I just want to make sure I'm understanding.

Reply:
I think I had a stroke or something or it is https://github.com/sailfishos/connman/pull/39 ... Yeah, there should be a negation there

Reply:
Changed with 1567662d64749df840efb4319ba357d435740a5d.